issuance/issuance.go

```diff
@@ -649,12 +658,30 @@ func (i *Issuer) Issue(req *IssuanceRequest) ([]byte, error) {
 
 	// check that the tbsCertificate is properly formed by signing it
 	// with a throwaway key and then linting it using zlint
-	err = i.Linter.Check(template, req.PublicKey)
+	lintCertBytes, err := i.Linter.Check(template, req.PublicKey)
 	if err != nil {
-		return nil, fmt.Errorf("tbsCertificate linting failed: %w", err)
+		return nil, nil, fmt.Errorf("tbsCertificate linting failed: %w", err)
+	}
+
+	return lintCertBytes, &issuanceToken{sync.Mutex{}, template, req.PublicKey}, nil
+}
+
+// Issue performs a real issuance using an issuanceToken resulting from a
+// previous call to Prepare(). Call this at most once per token. Calls after
+// the first will receive an error.
+func (i *Issuer) Issue(token *issuanceToken) ([]byte, error) {
+	if token == nil {
+		return nil, errors.New("nil issuanceToken")
+	}
+	token.Lock()
```

Comment:
Should the locking/unlocking come at the top of the function just in-case token is mutated in the nanoseconds between the call to Issue and the nil check happening?

Reply:
`token` is a local variable containing a pointer. The pointed-to data (e.g. token.template) can be modified concurrently, but we know the `token` variable itself cannot be modified concurrently because it is a local variable[1]. So checking for `token == nil` is safe without locking, but checking for `token.template == nil` is not safe without locking.

Further, it would be putting the cart before the horse to write:

```go
token.Lock()
defer token.Unlock()
if token == nil {
  return nil, errors.New("nil issuancetoken")
}
```

Because in the case that `token` is nil, we would already have panicked trying to dereference a nil pointer in that first line, `token.Lock()`.

[1]: And because there is no closure that has captured it.